<span style="font-family:Avenir;">
<font color="yellow"> <h1> Size </font> & Contribution Effect </h1> 
</span>

<span style="font-family:Avenir;">
<font color="LavenderBlush"> <h2> Stating Questions </h2> </font>
</span>



<span style="font-family:; font-size:1.4em;">
What is the <font color='yellow'> average </font> size of repositories over time for the top <font color='yellow'>3</font> programming languages in terms of pull requests?
</span>



<span style="font-family:;">

<table style="width:100%; border-collapse:collapse; table-layout:fixed;">
  <thead>
    <tr>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Set Expectations</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Collect Data</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Match Expectations & Data</font>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      This question is answerable and it's interesting.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      There is no existent answer for such question. It reflects the relationship between contriubtions and size of the repository.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Both match!
      </td>
    </tr>
  </tbody>
</table>

</span>



<span style="font-family:Avenir;">
<font color="LavenderBlush"> <h2> Exploratory Data Analytics</h2> </font>
</span>

# Loading the data

In [2]:
import sys; sys.path.append('../../')
from DataPreparation.Preprocess import read_data
from DataPreparation.Visualize import convey_insights

In [2]:
ds,_= read_data(split="all", fix=False, handle_langs=True, handle_useless="drop")

In [3]:
# Extract relevant columns: primaryLanguage, languagesSizes, pullRequests
ds = ds[['primaryLanguage', 'languagesUsed', 'languagesSizes','pullRequests']]

convey_insights([f"there are {ds['primaryLanguage'].nunique()} primaryLanguages"], "Number of primary programming languages")

<h3><font color="pink" size=5>Number of primary programming languages</font></h3> <font size=4>
<font color="pink">✦</font> there are 401 primaryLanguages<br><br></font>

<span style="font-family:;">

<table style="width:100%; border-collapse:collapse; table-layout:fixed;">
  <thead>
    <tr>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Set Expectations</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Collect Data</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Match Expectations & Data</font>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      I expect that there are more than 250 programming languages in the dataset.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Indeed, there are 401 different primary languages in the dataset.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Expectation agrees with data. 
      </td>
    </tr>
  </tbody>
</table>

</span>



<span style="font-family:Avenir;">
<font color="LavenderBlush"> <h2> Model Building</h2> </font>
</span>

In [4]:
import Logic
import importlib
importlib.reload(Logic)
pre_processed_ds = Logic.preprocess_languages_column(ds.copy())

In [18]:
import Logic
import importlib
importlib.reload(Logic)

top_languages = Logic.get_top_k_programming_languages(pre_processed_ds, k=3)

In [19]:
top_languages.head()

language  total pullRequests  average diskUsageKb
0  JavaScript             1149807           836.031047
1  TypeScript              985966           356.830300
2      Python              866438           404.479674

<span style="font-family:;">

<table style="width:100%; border-collapse:collapse; table-layout:fixed;">
  <thead>
    <tr>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Set Expectations</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Collect Data</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Match Expectations & Data</font>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      It's expected that the model will return the top 3 programming languages with their corresponding repositries' avg. sizes.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
        The top 3 programming languages with their corresponding repositries' avg. sizes are shown in the table above.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
        Results seems resonable.  
      </td>
    </tr>
  </tbody>
</table>

</span>



<span style="font-family:Avenir;">
<font color="LavenderBlush"> <h2> Results & Interpretation</h2> </font>
</span>

In [22]:
convey_insights([
    "JavaScript, TypeScript, and Python are the top 3 programming languages used in terms of pull requests",
    "Sizes are correlated with pull requests to some extent",
], "Results")

<h3><font color="pink" size=5>Results</font></h3> <font size=4>
<font color="pink">✦</font> JavaScript, TypeScript, and Python are the top 3 programming languages used in terms of pull requests<br><br><font color="pink">✦</font> Sizes are correlated with pull requests to some extent<br><br></font>

<span style="font-family:;">

<table style="width:100%; border-collapse:collapse; table-layout:fixed;">
  <thead>
    <tr>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Set Expectations</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Collect Data</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Match Expectations & Data</font>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      I expect JavaScript to be at least in the top 3 programming languages. As it's has a lot of usage in web development. 
      Also I expect to see a correlation between the number of pull requests and the size of the repository.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Data is shown above
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Expectation agrees with data.
      </td>
    </tr>
  </tbody>
</table>

</span>



<span style="font-family:Avenir;">
<font color="LavenderBlush"> <h2> Communicating Results</h2> </font>
</span>

In [3]:
### Probably code that will generate markdown
convey_insights([
    "The open source community is dominated by JavaScript, TypeScript, and Python",
    "TypeScript actually prevalence rate is promising as it is a relatively new language but it is already the second most used language in the open source community",
    "Sizes are correlated with pull requests to some extent",
], "Results")


<h3><font color="pink" size=5>Results</font></h3> <font size=4>
<font color="pink">✦</font> The open source community is dominated by JavaScript, TypeScript, and Python<br><br><font color="pink">✦</font> TypeScript actually prevalence rate is promising as it is a relatively new language but it is already the second most used language in the open source community<br><br><font color="pink">✦</font> Sizes are correlated with pull requests to some extent<br><br></font>

<span style="font-family:;">

<table style="width:100%; border-collapse:collapse; table-layout:fixed;">
  <thead>
    <tr>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Set Expectations</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Collect Data</font>
      </th>
      <th style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      <font color="DeepSkyBlue">Match Expectations & Data</font>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Professor & TA will appreciate and be interested in the communicated information.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      As above.
      </td>
      <td style="padding:10px; border:0.5px solid grey; text-align: center; width:33.33%;">
      Will be known after the presentation day.
      </td>
    </tr>
  </tbody>
</table>

</span>

